app/codefilter.cpp

```diff
@@ -0,0 +1,91 @@
+#include "codefilter.h"
+#include <QVideoFilterRunnable>
+#include <QFuture>
+
+#include <QtMultimedia/qvideoframe.h>
+#include <QImage>
+#include <QOpenGLContext>
+#include <QOpenGLFunctions>
+#include <QDebug>
+
+void processImage( QRDecoder *decoder, const QImage &image )
+{
+  decoder->process( image );
+};
+
+class QRRunnable : public QVideoFilterRunnable
+{
+
+  public:
+
+    QRRunnable( CodeFilter *filter )
+    {
+      mFilter = filter;
+    }
+
+    QVideoFrame run( QVideoFrame *input,
+                     const QVideoSurfaceFormat &surfaceFormat,
+                     QVideoFilterRunnable::RunFlags flags ) override
+    {
+      Q_UNUSED( surfaceFormat );
+      Q_UNUSED( flags );
+
+      if ( mFilter->decoder()->isDecoding() )
+      {
+        return *input;
+      }
+
+      if ( mFilter->futureThread().isRunning() )
+      {
+        return *input;
+      }
+
+      mFilter->decoder()->setVideoFrame( *input );
+
+      QImage captured = QRDecoder::videoFrameToImage( *input );
+      mFilter->futureThread() = QtConcurrent::run( processImage, mFilter->decoder(), captured );
+
+      return *input;
+    }
+
+  private:
+    CodeFilter *mFilter;
+};
+
+CodeFilter::CodeFilter()
+{
+  mDecoder = new QRDecoder;
+
+  QObject::connect( mDecoder, &QRDecoder::capturedChanged, this, &CodeFilter::setCapturedData );
+}
+
+QVideoFilterRunnable *CodeFilter::createFilterRunnable()
+{
+  return new QRRunnable( this );
```

Comment:
We should probably save this address to a new attribute and free it in `CodeFilter`'s destructor or somehow destroy QRRunnable instance when `this` (CodeFilter) is deleted

Reply:
From documentation, it seems that destructions is handled. Also added docs to the function.
http://doc.qt.io/qt-5/qabstractvideofilter.html#createFilterRunnable